qiskit/primitives/containers/data_bin.py

```diff
@@ -0,0 +1,89 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Dataclass tools for data namespaces (bins)
+"""
+from __future__ import annotations
+
+from dataclasses import make_dataclass
+from typing import Iterable, Tuple
+
+
+class DataBinMeta(type):
+    """Metaclass for :class:`DataBin` that adds the shape to the type name.
+
+    This is so that the class has a custom repr with DataBin<*shape> notation.
+    """
+
+    def __repr__(cls):
+        name = cls.__name__
+        if cls._SHAPE is None:
+            return name
+        shape = ",".join(map(str, cls._SHAPE))
+        return f"{name}<{shape}>"
+
+
+class DataBin(metaclass=DataBinMeta):
+    """Base class for data bin containers.
+
+    Subclasses are typically made via :class:`~make_data_bin`, which is a specialization of
+    :class:`make_dataclass`.
+    """
+
+    _RESTRICTED_NAMES = ("_RESTRICTED_NAMES", "_SHAPE", "_FIELDS", "_FIELD_TYPES")
+    _SHAPE: Tuple[int, ...] | None = None
+    _FIELDS: Tuple[str, ...] = ()
+    """The fields allowed in this data bin."""
+    _FIELD_TYPES: Tuple[type, ...] = ()
+    """The types of each field."""
+
+    def __repr__(self):
+        vals = (f"{name}={getattr(self, name)}" for name in self._FIELDS if hasattr(self, name))
+        return f"{type(self)}({', '.join(vals)})"
+
+
+def make_data_bin(
+    fields: Iterable[Tuple[str, type]], shape: Tuple[int, ...] | None = None
```

Comment:
If we are going all in on 3.9 python types we should not use List, Tuple, Dict, but the python classes themselves
```suggestion
    fields: Iterable[tuple[str, type]], shape: tuple[int, ...] | None = None
```
(and same elsewhere to be consistent)

Reply:
I updated the other parts, but forgot about this part. Thanks.